Opentrons/opentrons
# feat(robot-server): Executor for protocol session

# Overview

This PR is a step towards running protocols via sessions API and not RPC. 

This PR fills out the ProtocolSession with command execution. Features supported:
- Protocol sessions support session commands `startRun, startSimulate, pause, resume, and cancel`
- support python files can be imported using `import {name}`
- support files can be opened using `open({name})`

What this PR **does not** address:
- The schema for the protocol session detailed response. Specifically the command history and errors. 
- Extra labware arg to Session.

# Changelog

Added three classes: 

`ProtocolCommandExecutor` - This class is derived from `CommandExecutor`. Its jobs:
- To create a `_Worker` and `ProtocolRunner`
-  to validate whether a command can be executed in the current state.
- to delegate commands to a `Worker`
- to listen for command from a `ProtocolRunner`

`ProtocolRunner` - This class wraps a `opentrons.api.session.Session` and `Broker`. It synchronously loads, simulates, and runs protocol. Listeners can be provided to monitor the flow of the protocol (basically just creating a Broker and dispatching the `COMMAND` and `SESSION` topics.)

`Worker`- This manages a `ProtocolRunner` in response to commands received via `ProtocolCommandExecutor`. An asyncio task runs in the background to process protocol jobs from commands (run, simulate, exit). A `ProtocolRunner` listener is used to monitor the flow and state of the protocol. Calls to `ProtocolRunner` are run in a separate thread `using loop.run_in_executor`.

# Review requests

This was tested with local dev robot-server. Would be lovely if it were tested on a robot.

# How to test

Assuming a protocol named "myproto.py" and a robot at `localhost`

- Upload a protocol `curl -X POST "http://localhost:31950/protocols" -H  "accept: application/json" -H  "Content-Type: multipart/form-data" -F "protocol_file=@myproto.py"`
- Create the protocol session: `curl -X POST "http://localhost:31950/sessions" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"data\":{\"type\":\"Session\",\"attributes\":{\"sessionType\":\"protocol\",\"createParams\":{\"protocolId\":\"myproto\"}}}}"`
- Extract the session id from response. For example, the UUID in this example is the session id.  `{"data":{"id":"98dc5beb-d800-4afc-8b74-85adb7c8358c","....`. Copy the id for later HTTP requests.
- Get results of session using `curl "http://localhost:31950/sessions98dc5beb-d800-4afc-8b74-85adb7c8358c"`. The currentState should be `loaded` and all the simulated protocol steps should be a number of commands in the `commands` list.
- Run protocol using `curl -X POST "http://localhost:31950/sessions/98dc5beb-d800-4afc-8b74-85adb7c8358c/commands/execute" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"data\":{\"type\":\"SessionCommand\",\"attributes\":{\"command\":\"protocol.startRun\",\"data\":{}}}}"`
- Use the same format for issuing other commands: `protocol.startSimulate, protocol.pause, protocol.resume, protocol.cancel`
- Delete session `curl -X DELETE http://localhost:31950/sessions98dc5beb-d800-4afc-8b74-85adb7c8358c`
- Delete the protocol `curl -X DELETE "http://localhost:31950/protocols/myproto`


# Risk assessment

Very little. The RunApp does not use this API. This feature will initially be released for "power" users as HTTP only.


closes #6192 
closes #6194 
closes #6195 
closes #6196
